### PR TITLE
Expose Setting for prefetch thresholds

### PR DESCRIPTION
**Pull request by tadgh**
Related to discussion here: https://github.com/hapifhir/hapi-fhir/issues/6109 

- expose prefetch threshold setting
- add docs

@bashir2

### CODE REVIEW

File: src/main/resources/application.yaml [hunk 1]
```diff
@@ -76,7 +76,7 @@ hapi:
     ### This flag when enabled to true, will avail evaluate measure operations from CR Module.
     ### Flag is false by default, can be passed as command line argument to override.
     cr:
-      enabled: false
+      enabled: true
```
Comment (codeforgreen) [inline src/main/resources/application.yaml:79]: **question**: is this an intentional change? I didn't see how it relates to the reported issue.

**Review by codeforgreen — APPROVED** [review]

File: src/main/resources/application.yaml (hunk 1 above)
Comment (tadgh) [reply src/main/resources/application.yaml:79]: ah good catch, not intentional

File: src/main/resources/application.yaml [hunk 2]
```diff
@@ -229,7 +229,18 @@ hapi:
...
+    # Search Prefetch Thresholds.
+
+    # This setting sets the number of search results to prefetch. For example, if this list
+    # is set to [100, 1000, -1] then the server will initially load 100 results and not
+    # attempt to load more. If the user requests subsequent page(s) of results and goes
+    # past 100 results, the system will load the next 900 (up to the following threshold of 1000).
+    # The system will progressively work through these thresholds.
+    # A threshold of -1 means to load all results. Note that if the final threshold is a
+    # number other than -1, the system will never prefetch more than the given number.
+    search_prefetch_thresholds: 13,503,2003
```
Comment (bashir2) [inline src/main/resources/application.yaml:242]: Based on the above comment, shouldn't this list end with `-1` (similar to the [default](https://github.com/hapifhir/hapi-fhir/blob/20d3e6bb252eff52def255669ca6f590c2d8db59/hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/config/JpaStorageSettings.java#L191))?

Comment (tadgh) [reply src/main/resources/application.yaml:242]: Ah thank you, yes, you are correct. patched

**Comment by tadgh** [comment]
@bashir2 this look better?

File: src/main/java/ca/uhn/fhir/jpa/starter/AppProperties.java [hunk 3]
```diff
@@ -104,6 +104,8 @@ public class AppProperties {
 
 	private final List<String> custom_provider_classes = new ArrayList<>();
 
+	private List<Integer> search_prefetch_thresholds = List.of(13, 503, 2003, -1);
```
Comment (bashir2) [inline src/main/java/ca/uhn/fhir/jpa/starter/AppProperties.java:107]: nit: Another option is not to call `jpaStorageSettings.setSearchPreFetchThresholds` if the list is not provided. I am just thinking about the edge-case where this goes out of sync from the the actual default [here](https://github.com/hapifhir/hapi-fhir/blob/20d3e6bb252eff52def255669ca6f590c2d8db59/hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/config/JpaStorageSettings.java#L191).

**Review by bashir2 — APPROVED** [review]
Thanks @tadgh for the quick fix.

File: src/main/java/ca/uhn/fhir/jpa/starter/AppProperties.java (hunk 3 above)
Comment (tadgh) [reply src/main/java/ca/uhn/fhir/jpa/starter/AppProperties.java]: Good call! I have patched it.

File: src/main/resources/application.yaml [hunk 4]
```diff
@@ -229,7 +229,18 @@ hapi:
...
+    # Search Prefetch Thresholds.
+
+    # This setting sets the number of search results to prefetch. For example, if this list
+    # is set to [100, 1000, -1] then the server will initially load 100 results and not
+    # attempt to load more. If the user requests subsequent page(s) of results and goes
+    # past 100 results, the system will load the next 900 (up to the following threshold of 1000).
+    # The system will progressively work through these thresholds.
+    # A threshold of -1 means to load all results. Note that if the final threshold is a
+    # number other than -1, the system will never prefetch more than the given number.
+    search_prefetch_thresholds: 13,503,2003,-1
```
Comment (jkiddo) [inline src/main/resources/application.yaml:242]: If these are the default values shouldn't they be provided as such in the code here https://github.com/hapifhir/hapi-fhir-jpaserver-starter/pull/742/files#diff-79251bbeac3bb7f417a707bbda464fdfb2394f40bd88e43a9f5472e0ed351c83R107?

Comment (tadgh) [reply src/main/resources/application.yaml:242]: Nah its fine, this is only populated if non-empty, and will rely on the HAPI-FHIR defaults if absent/empty. I put these in the properties file here for discoverability only. I hesitate to make them a java default in jpaserver-starter for fear of drift between hapi and this project.

**Review by dotasek — APPROVED** [review]
Looks good to me. Tracing startup shows the settings being set via jpaStorageSettings.setSearchPreFetchThresholds in FhirServerConfigCommon

Can confirm what @tadgh was saying with regard to defaults; if the setting is empty in application.yaml, the default values from HAPI FHIR are used (link below)

https://github.com/hapifhir/hapi-fhir/blob/71cb097aa647657e7073990dc16d90d566020bb4/hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/config/JpaStorageSettings.java#L192

**Review by jkiddo — APPROVED** [review]

**Pull request merged**